src/com/trollworks/gcs/character/TextTemplate.java

```diff
@@ -574,19 +644,37 @@ private void emitKey(BufferedReader in, BufferedWriter out, String key, File bas
                 processSpellsLoop(out, extractUpToMarker(in, KEY_SPELLS_LOOP_END));
             } else if (key.startsWith(KEY_MELEE_LOOP_START)) {
                 processMeleeLoop(out, extractUpToMarker(in, KEY_MELEE_LOOP_END));
+            } else if (key.startsWith(KEY_HIERARCHICAL_MELEE_LOOP_START)) {
+                processHierarchicalMeleeLoop(out, extractUpToMarker(in, KEY_HIERARCHICAL_MELEE_LOOP_END));
             } else if (key.startsWith(KEY_RANGED_LOOP_START)) {
                 processRangedLoop(out, extractUpToMarker(in, KEY_RANGED_LOOP_END));
+            } else if (key.startsWith(KEY_HIERARCHICAL_RANGED_LOOP_START)) {
+                processHierarchicalRangedLoop(out, extractUpToMarker(in, KEY_HIERARCHICAL_RANGED_LOOP_END));
             } else if (key.startsWith(KEY_EQUIPMENT_LOOP_START)) {
                 processEquipmentLoop(out, extractUpToMarker(in, KEY_EQUIPMENT_LOOP_END));
             } else if (key.startsWith(KEY_NOTES_LOOP_START)) {
                 processNotesLoop(out, extractUpToMarker(in, KEY_NOTES_LOOP_END));
+            } else if (key.startsWith(KEY_ONLY_CATEGORIES)) {
+                setOnlyCategories(key);
+            } else if (key.startsWith(KEY_EXCLUDE_CATEGORIES)) {
+                setExcludeCategories(key);
             } else {
                 writeEncodedText(out, String.format(UNIDENTIFIED_KEY, key));
             }
             break;
         }
     }
 
+    private void setOnlyCategories(String key) {
+        String[] categories = key.substring(KEY_ONLY_CATEGORIES.length()).split(UNDERSCORE);
+        mOnlyCategories.addAll(Arrays.asList(categories));
```

Comment:
Perhaps `mOnlyCategories` and `mExcludedCategories` (below) should be emptied before setting the values here? Otherwise, you can only ever add to the lists.

I see later on that you actually implicitly clear the values at certain points... I guess that's reasonable, too.

Reply:
Yes, there were a few ways I could do it, so I chose to allow the template to add to either list using @ONLY_CATEGORIES_ and @EXCLUDE_CATEGORIES_ prior to executing any loop.   Either on one line, or multiple.